File: tools/babypsyduck/previews.js
// Renders the preview frames used by previews.py (docs/babypsyduck):
//   out/gif/f###.png   animation grid, one PNG per GIF frame
//   out/prev_*.png     variants, turnaround, expressions
'use strict';
const fs = require('fs');
const path = require('path');
const { render, close } = require('./render');

const ID = 'babypsyduck';
const CELLS = [['ground_idle', true], ['ground_walk', true], ['battle_idle', true], ['sleep', true],
  ['water_idle', true], ['water_swim', true], ['headache', false], ['cry', false],
  ['physical', false], ['special', false], ['recoil', false], ['faint', false]];
const COLS = 4, TILE = 200, CYCLE = 4.0, FPS = 12.5;
const out = (f) => path.join(__dirname, 'out', f);

(async () => {
  const bb = JSON.parse(fs.readFileSync(out(`${ID}/${ID}.bbmodel`), 'utf8'));
  const lengths = Object.fromEntries(bb.animations.map((a) => [a.name.split('.').pop(), a.length]));
  // static images
  const v = { yaw: 28, pitch: 14, anim: 'ground_idle', time: 0.4 };
  await render(bb, out('prev_normal.png'), { frames: [v], cols: 1, tile: 360, ground: false, bg: 0x2b2d3a });
  await render(bb, out('prev_shiny.png'), { frames: [v], cols: 1, tile: 360, texture: 1, ground: false, bg: 0x2b2d3a });
  // alpha: glowing eyes layer drawn over the base texture, in a dark scene so the glow reads
  await render(bb, out('prev_alpha.png'), { frames: [v], cols: 1, tile: 360, emissive: [2], ambient: 0.35, diffuse: 0.35, ground: false, bg: 0x16171f });
  await render(bb, out('prev_views.png'), { frames: [0, 45, 90, 135, 180, 270].map((yaw) => ({ yaw, pitch: 12 })), cols: 6, tile: 240, ground: false, bg: 0x2b2d3a });
  await render(bb, out('prev_views_shiny.png'), { frames: [0, 45, 90, 135, 180, 270].map((yaw) => ({ yaw, pitch: 12 })), cols: 6, tile: 240, texture: 1, ground: false, bg: 0x2b2d3a });
  const face = (anim, time) => ({ yaw: 10, pitch: 4, anim, time });
  await render(bb, out('prev_expr.png'), {
    frames: [face(null, 0), face('blink', 0.1), face('cry', 0.35), face('headache', 1.0), face('faint', 0.5)],
    cols: 5, tile: 240, ground: false, bg: 0x2b2d3a, center: [0, 16, -6], radius: 11,
  });
  // animation grid frames
  fs.mkdirSync(out('gif'), { recursive: true });
  const n = Math.round(CYCLE * FPS);
  for (let k = 0; k < n; k++) {
    const t = k / FPS;
    const frames = CELLS.map(([anim, loop]) => {
      const len = lengths[anim];
      // one-shots play once, hold the last pose, then restart with the loop
      return { yaw: 28, pitch: 14, anim, time: loop ? t : Math.min(t, len) };
    });
    await render(bb, out(`gif/f${String(k).padStart(3, '0')}.png`), { frames, cols: COLS, tile: TILE, radius: 19, center: [0, 11, 0] });
  }
  fs.writeFileSync(out('gif/labels.json'), JSON.stringify(CELLS.map((c) => c[0])));
  await close();
})();

File: tools/babypsyduck/previews.py
# Assembles docs/babypsyduck images from the renders made by previews.js.
import glob, json, os
from PIL import Image, ImageDraw, ImageFont

HERE = os.path.dirname(os.path.abspath(__file__))
OUT = os.path.join(HERE, 'out')
DOCS = os.path.join(HERE, '..', '..', 'docs', 'babypsyduck')
os.makedirs(DOCS, exist_ok=True)

def font(size, bold=True):
    try:
        return ImageFont.truetype('/usr/share/fonts/truetype/dejavu/DejaVuSans%s.ttf' % ('-Bold' if bold else ''), size)
    except Exception:
        return ImageFont.load_default()

def label(dr, xy, text, size=15, fill=(255, 255, 255)):
    dr.text(xy, text, fill=fill, font=font(size), stroke_width=2, stroke_fill=(40, 40, 50))

# animation grid gif
labels = json.load(open(os.path.join(OUT, 'gif', 'labels.json')))
TILE, COLS = 200, 4
frames = []
for f in sorted(glob.glob(os.path.join(OUT, 'gif', 'f*.png'))):
    im = Image.open(f).convert('RGB')
    dr = ImageDraw.Draw(im)
    for i, lab in enumerate(labels):
        x, y = (i % COLS) * TILE, (i // COLS) * TILE
        dr.rectangle((x, y, x + TILE - 1, y + TILE - 1), outline=(70, 70, 84))
        label(dr, (x + 7, y + 5), lab, 14)
    frames.append(im.quantize(colors=128, method=Image.Quantize.MEDIANCUT, dither=Image.Dither.NONE))
frames[0].save(os.path.join(DOCS, 'animations.gif'), save_all=True, append_images=frames[1:], duration=80, loop=0, optimize=True)

# variants: normal / shiny / alpha
names = [('normal', 'Normal'), ('shiny', 'Shiny'), ('alpha', 'Alpha (glowing eyes)')]
o = Image.new('RGB', (360 * 3, 360), (43, 45, 58))
d = ImageDraw.Draw(o)
for i, (n, t) in enumerate(names):
    o.paste(Image.open(os.path.join(OUT, 'prev_%s.png' % n)), (i * 360, 0))
    label(d, (i * 360 + 12, 10), t, 18, (255, 214, 102) if n == 'shiny' else (255, 255, 255))
o.save(os.path.join(DOCS, 'variants.png'), optimize=True)

# turnaround (normal + shiny)
a = Image.open(os.path.join(OUT, 'prev_views.png')); b = Image.open(os.path.join(OUT, 'prev_views_shiny.png'))
o = Image.new('RGB', (a.width, a.height * 2), (43, 45, 58))
o.paste(a, (0, 0)); o.paste(b, (0, a.height))
o.save(os.path.join(DOCS, 'views.png'), optimize=True)

# expressions
e = Image.open(os.path.join(OUT, 'prev_expr.png')).convert('RGB')
d = ImageDraw.Draw(e)
for i, t in enumerate(['open', 'blink / sleep', 'cry (happy)', 'headache', 'faint (dizzy)']):
    label(d, (i * 240 + 10, 8), t, 15)
e.save(os.path.join(DOCS, 'expressions.png'), optimize=True)
print('wrote', sorted(os.listdir(DOCS)))
